vlog4j-core/src/main/java/org/semanticweb/vlog4j/core/reasoner/implementation/VLogReasoner.java

```diff
@@ -808,4 +812,46 @@ ReasonerState getReasonerState() {
 	void setReasonerState(ReasonerState reasonerState) {
 		this.reasonerState = reasonerState;
 	}
+
+	@Override
+	public void writeInferences(OutputStream stream) throws IOException {
+		Set<Predicate> toBeQueriedHeadPredicates = new HashSet<Predicate>();
+		for (Rule rule : this.knowledgeBase.getRules()) {
+			for (Literal literal : rule.getHead()) {
+				toBeQueriedHeadPredicates.add(literal.getPredicate());
+			}
+		}
+		for (DataSourceDeclaration dataSourceDeclaration : this.knowledgeBase.getDataSourceDeclarations()) {
+			toBeQueriedHeadPredicates.add(dataSourceDeclaration.getPredicate());
+		}
+		for (Fact fact : this.knowledgeBase.getFacts()) {
+			toBeQueriedHeadPredicates.add(fact.getPredicate());
+		}
+
+		for (Predicate predicate : toBeQueriedHeadPredicates) {
+			List<Term> tobeGroundedVariables = new ArrayList<Term>();
+			for (int i = 0; i < predicate.getArity(); i++) {
+				tobeGroundedVariables.add(Expressions.makeUniversalVariable("X" + i));
+			}
+			try (final QueryResultIterator answers = this
+					.answerQuery(Expressions.makePositiveLiteral(predicate, tobeGroundedVariables), true)) {

```

Comment:
We should not call answerQuery directly, as it creates too many objects (for each `TermQueryResultIterator`, it creates a `VLogQueryResultIterator` to include the `Correctness` result, but for  `writeInferedFacts`, the `Correctness` result is the same for all querries. 
Also, logWarningOnCorrectness() is called for every querried predicate. If `Correctness `is not `SOUND_AND_COMPLETE` and WARNING logging is enabled, a warning will be logged for each queried predicate, which is both slow and annoying.

Also, `answerQuery` does checks for each query that only need to be done once (reasoner not closed, kb loaded) or never (query not null)  for  `writeInferedFacts` .